src/federation.cpp

```diff
@@ -64,13 +75,172 @@ static std::shared_ptr<CWallet> GetWalletForBlockSigning()
 }
 #endif // ENABLE_WALLET
 
+//! All public keys contained within the block signing script.
+static std::vector<CPubKey> g_block_signing_keys;
+//! The block signing public key associated with this node.
+static CPubKey g_our_block_signing_key;
+//! The index of the block signing key associated with this node.
+static int g_our_block_signing_key_index = 0;
+
+//! The subnets match only the IPs which are known federation block-signing
+//! nodes.  Block-signing messages from other nodes not matching any of these
+//! subnets are ignored, so as to not reveal that we are a block-signing
+//! federation node.
+static std::set<CSubNet> g_whitelist;
+
+struct BlockSigner {
+    CPubKey m_pubkey;
+    std::set<std::string> netaddrs;
+    int64_t m_endpoint_last_update_time;
+    std::set<NodeId> m_nodes;
+
+    BlockSigner() : m_endpoint_last_update_time(0) { }
+    BlockSigner(const CPubKey& pubkey, const std::string& netaddr) : m_pubkey(pubkey), netaddrs({netaddr}), m_endpoint_last_update_time(0) { }
+};
+
+//! A record to store configuration information and state for each block signer.
+static std::map<CPubKey, BlockSigner> g_block_signers;
+
+// Parses the configuration options to determine connection and pubkey
+// information for all the other federation peers.
+static bool ReadFederationOpts() {
+    using std::swap;
+
+    LOCK(cs_block_signer);
+
+    CScript script = Params().GetConsensus().signblockscript;
+#if ENABLE_WALLET
+    std::shared_ptr<CWallet> pwallet = GetWalletForBlockSigning();
+    LegacyScriptPubKeyMan* spk_man = pwallet ? pwallet->GetOrCreateLegacyScriptPubKeyMan() : nullptr;
+#endif // ENABLE_WALLET
+    CScript::const_iterator pc = script.begin();
+    opcodetype opcode;
+    std::vector<unsigned char> data;
+    while (pc < script.end() && script.GetOp(pc, opcode, data)) {
+        // We are only interested in data pushes
+        if (data.empty()) {
+            continue;
+        }
+        // Check that the push data is a valid pubkey
+        CPubKey pk(data);
+        if (!pk.IsFullyValid()) {
+            continue;
+        }
+        // Treat any valid pubkey as a block-signing key
+        g_block_signing_keys.push_back(pk);
+#if ENABLE_WALLET
+        // And check to see if it is *our* block-signing key
+        if (spk_man && spk_man->HaveKey(pk.GetID())) {
+            g_our_block_signing_key = pk;
+        }
+#endif // ENABLE_WALLET
+    }
+
+    if (!g_our_block_signing_key.IsValid()) {
+        LogPrintf("Error: unable to determine block-signing key.\n");
+        return false;
+    }
+
+    std::sort(g_block_signing_keys.begin(),
+              g_block_signing_keys.end());
+
+    g_our_block_signing_key_index = 0;
```

Comment:
That's precisely what I had in mind for production: a pubkey set as a configuration option, with the privkey stored within a secure hardware enclave or attached hardware wallet. I haven't really worked out the low-level details of this though. For example, do we keep the wallet enabled and add the pubkey as a watch-only key? Or do we remove the wallet and replicate the features we need for a smaller exposed attack surface? Honestly I haven't spend more than 5 minutes thinking about that yet. Obviously right now if you try to run this without the wallet enabled it will error out because it can't find any keys it recognizes.

Reply:
I moved the wallet-check into the post-sorted loop as suggested because I think that's cleaner. Thanks for the suggestion. I'm not marking this as resolved yet though because of the other unresolved aspects.